File: src/Day9/Day9-2.py
file_path9 = './src/Day9/data9.txt'

with open(file_path9, 'r', encoding='utf-8') as file:
    file_content9 = file.read()

# Split the content into lines
lined_data9 = file_content9.split('\n')

data_array = map(lambda x: x.split(' '), lined_data9)
squared_numbers_list = list(data_array)

array_of_arrays_of_integers = [[int(element) for element in inner_array] for inner_array in squared_numbers_list]


# print(array_of_arrays_of_integers)

testdata9 = [
    [0, 3, 6, 9, 12, 15],
    [1, 3, 6, 10, 15, 21],
    [10, 13, 16, 21, 30, 45]
]

all_all_differences =  []

for array in array_of_arrays_of_integers:
    all_differences = [array]

    def find_differences(array):
        differences = [array[i+1] - array[i] for i in range(len(array)-1)]
        all_differences.append(differences)
        if all(diff == 0 for diff in differences):
            return all_differences
        else:
            return find_differences(differences)

    all_all_differences.append(find_differences(array))

print(all_all_differences)


def find_new_end(input_array):
    final_value = 0

    input_array[-1].append(0)

    for i in range(len(input_array)-2, -1, -1):
        new_value = input_array[i][0] - input_array[i+1][0]
        if i == 0:
            final_value = new_value
        input_array[i].insert(0, new_value)
    print(final_value)
    return final_value


accumulator = 0

for i in range(len(array_of_arrays_of_integers)):
    accumulator += find_new_end(all_all_differences[i])

print(accumulator)

File: src/Day9/Day9-2.ts
let fs9_2 = require('fs');


// Read the content of the file
const filePath9_2 = './src/Day9/data9.txt';
const fileContent9_2 = fs9_2.readFileSync(filePath9_2, 'utf-8'); // this reads the file synchronously

const lined_data9_2: string[]  = fileContent9_2.split('\n');

let data9_2_array: string[][] = []
lined_data9_2.map(el => data9_2_array.push(el.split(' ')))
let data9_2_array_numbers: number[][] = data9_2_array.map(el => el.map(el2 => Number(el2)))

const testdata9_2: number[][] = [
[0, 3, 6, 9, 12, 15],
[1, 3, 6, 10, 15, 21],
[10, 13, 16, 21, 30, 45]
]

let all_all_differences_2: number[][][] = []

for (let i = 0; i < data9_2_array_numbers.length; i++) {

    let all_differences: number[][] = [data9_2_array_numbers[i]];

    function findDifferences(array: number[]): number[][] {
        let differences: number[] = [];
        for (let i = 0; i < array.length - 1; i++) {
            differences.push(array[i+1] - array[i])
        }
        all_differences.push(differences)
        if (differences.every(el => el === 0)) {
            return all_differences
        } else {
            findDifferences(differences)
        }

        return all_differences
    }

    all_all_differences_2.push(findDifferences(data9_2_array_numbers[i]))
}

console.log(all_all_differences_2)

function findNewEnd_2(input: number[][]): number {
    let final_value: number = 0 

    input[input.length - 1].unshift(0)

    for (let i = input.length - 2; i >= 0; i--) { // add the two ends and push it on to the array
        console.log(i)

        let length_current: number = input[i].length;
        // console.log('length_current', length_current)
        console.log('current start', input[i][0])
        console.log('previous start', input[i+1][0])
        let new_value: number = input[i][0] - input[i+1][0]
        if (i === 0) {
            final_value = new_value
        }
        console.log('new value', new_value)
        input[i].unshift(new_value)
    }

    return final_value

}

// console.log(findNewEnd_2(all_all_differences_2[0]))

let accululator_2: number = 0

for (let i = 0; i < data9_2_array_numbers.length; i++) {
    accululator_2 += findNewEnd_2(all_all_differences_2[i])
}

console.log(accululator_2)